### PR TITLE
Simple ptrace probe that reports writes to stdout and stderr

### PR DESCRIPTION
**Pull request by PabloMansanet**
image

I decided to go with the `stdout` and `stderr` writes as an example as they have an easy to interpret first argument and return value. It's probably not ideal to go much further with `ptrace` because of the difficulty of tracing a multithreaded program that forked before attaching the probe. This probe simply traces the `main` thread.

Oh, it also should be run with `root` permissions; it's difficult to safely allow ptrace to be used by a non-root user.

### CODE REVIEW

File: crates/ptrace_probe/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+fn handle_syscall(pid: Pid) -> Result<Option<Event>, color_eyre::Report> {
+    ptrace::syscall(pid, None)?;
+    waitpid(Some(pid), None)?;
+    let registers = ptrace::getregs(pid)?;
+    let syscall: Sysno = (registers.orig_rax as i32).into();
+    let rdi = registers.rdi;
+
+    // At this point, the syscall is suspended, we need to call `syscall`
+    // again to actually execute it and retrieve the return value from `rax`.
```
Comment (goodhoko) [inline crates/ptrace_probe/src/main.rs:48]: Such a nice explanation. 🖤

File: crates/ptrace_probe/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+fn main() -> Result<()> {
+    color_eyre::install()?;
+    let args = Args::parse();
+    let pid = Pid::from_raw(args.pid as i32);
+
+    let client = args
+        .address
+        .map(Client::new)
+        .unwrap_or_else(Client::try_default)?;
```
Comment (goodhoko) [inline crates/ptrace_probe/src/main.rs]: I tend to like these functional-style pipeline-like constructs, but here (and it may be just my lack of rust proficiency) I find it confusing. It looks like that when `Client::new` fails, we fallback to `Client::try_default` but we definitely want to abort in that case.

Only after opening this in IDE with type hints I realized it works properly. 

I find what @skywhale [used](https://github.com/tonarino/acoustic_profiler/pull/22/files#diff-1e4ce9864165105480b1843623dfa14a41c6731ccc44584283d385a75b1776b7R21-R24) in his probe more readable:

```suggestion
    let client = match args.address {
        Some(address) => Client::new(address)?,
        None => Client::try_default()?,
    };
```

**Review by goodhoko — APPROVED** [review]
Looks good! I only have a single stylistic suggestion.

Unfortunately, I can't test this on my machine, but maybe @strohel could give it a try. 😉 

> It's probably not ideal to go much further with ptrace because of the difficulty of tracing a multithreaded program that forked before attaching the probe. This probe simply traces the main thread.

That's a bummer. Another nuisance is that ptrace allegedly[1] degrades performance of the traced program. Let's merge this anyway and keep it as a tool in our probe-belt. .)


[1] The second to last page of this awesome [zine](https://jvns.ca/strace-zine-unfolded.pdf).

**Comment by strohel** [comment]
> Unfortunately, I can't test this on my machine, but maybe @strohel could give it a try. :wink:

Will do!

> [1] The second to last page of this awesome [zine](https://jvns.ca/strace-zine-unfolded.pdf).

Oh that is pure love, how have you found this gem?

**Comment by PabloMansanet** [comment]
> [1] The second to last page of this awesome [zine](https://jvns.ca/strace-zine-unfolded.pdf).

Haha that's awesome! She has a lot on that blog, wow. Reminds me of a youtuber called [Vihart](https://www.youtube.com/watch?v=TINfzxSnnIE) that explained math concepts very creatively on a notebook.

And yeah, `ptrace` is very invasive so we should probably not use it on tonari proper.

File: crates/ptrace_probe/src/main.rs (hunk 2 above)
Comment (PabloMansanet) [reply crates/ptrace_probe/src/main.rs]: Agreed :) to be honest, even though I use these functional chains a lot, this one is pretty hard to follow, and I also had to double check to make sure the unwrapping order was correct as I was writing it.

Will go with the match alternative.

**Comment by goodhoko** [comment]
@PabloMansanet @strohel IDK, I stumbled upon Julia's zines when researching ptrace and dtrace. The one I linked is quite old and raw. Since then she produced quite a collection of polished zines at https://wizardzines.com/. They all look cool but I hadn't yet bought any.

I love vihart too, especially the videos on hexaflexagons!

File: Cargo.toml [hunk 3]
```diff
@@ -4,13 +4,15 @@ members = [
...
     "crates/test_probe",
+    "crates/ptrace_probe",
 ]
 
 # Specify a subset of member crates that compile on all supported architectures.
 default-members = [
     "crates/composer_api",
     "crates/composer",
     "crates/test_probe",
+    "crates/ptrace_probe",
```
Comment (strohel) [inline Cargo.toml:15]: @goodhoko / @skywhale this does compile on a Mac, right?

**Review by strohel — COMMENTED** [review]
Nice, I've tested (on a `cat`) and this indeed works. One curiosity question.

File: Cargo.toml (hunk 3 above)
Comment (strohel) [reply Cargo.toml:15]: @PabloMansanet oh I see you're removing this in a93fe06434693fdcb0c66d1c130f3016a325017e (not yet submitted in a PR).

Edit: ehm, I see that's @goodhoko who's branch/commit that is.

Comment (goodhoko) [reply Cargo.toml:15]: It doesn't. I only noticed this after merging. I do have a WIP branch where I remove this line and adapt the CI to  run both on linux and macOS to lint/test all crates.
